lib/rules/curly.js

```diff
@@ -309,6 +331,103 @@ module.exports = {
                 hasUnsafeIf(statement) && isFollowedByElseKeyword(node);
         }
 
+        /**
+         * Missing curly brace reporting location examples:
+         *
+         * The following location should be reported
+         *
+         * if (foo) bar()
+         *        ^^^
+         *
+         * while(foo) bar();
+         *          ^^^
+         *
+         * do foo() while(bar)
+         * ^^^
+         *
+         * for(;;) bar();
+         *       ^^^
+         *
+         * for (var foo of bar) console.log(foo)
+         *                    ^^^
+         *
+         * for (var foo in bar) console.log(foo)
+         *                    ^^^
+         *
+         * if (foo) {bar()} else bar(1);
+         *                  ^^^^
+         *
+         * Unexpected curly brace reporting location examples:
+         *
+         * The following location should be reported
+         *
+         * if (true) foo(); else { baz(); }
+         *                       ^
+         *
+         * for (;;) { foo(); }
+         *          ^
+         *
+         * while (bar) { foo(); }
+         *             ^
+         *
+         * do{foo();} while(bar);
+         *   ^
+         *
+         * for (var foo of bar) {console.log(foo)}
+         *                      ^
+         *
+         * if (foo) bar() else {bar(1)}
+         *                ^^^^
+         *
+         * Get the location of missing or non missing curly brace.
+         * @param {ASTNode} node the node to check.
+         * @param {string} name name to check.
+         * @param {boolean} isMissing is curly brace missing.
+         * @returns {Object} Position object.
+         */
+        function getReportLoc(node, name, isMissing) {
+            if (name === "else") {
+                if (!isMissing) {
+                    const token = sourceCode.getTokenAfter(getElseKeyword(node));
+
+                    return token.loc;
+                }
+
+                return getElseKeyword(node).loc;
+            }
+
+            if (blockStartLoc.has(name)) {
+                if (isMissing) {
+                    const token = sourceCode.getTokenBefore(node.body);
+
+                    return {
+                        start: token.loc.start,
+                        end: node.body.loc.start
+                    };
+                }
+
+                return createNextLocation(node.body.loc.start);
+            }
+
+            if (name === "if") {
+                return isMissing
+                    ? {
+                        start: node.test.loc.end,
```

Comment:
This could also underline whitespace between the `test` node and the opening paren:

![image](https://user-images.githubusercontent.com/44349756/88963880-cdf5d400-d2a8-11ea-81de-e474b66f47ed.png)

(the same applies to `while`)

Could we maybe unify all cases and thus simplify the code? I think the only difference is whether the "body" is `.body`, `.consequent`, or `.alternate` child, so maybe we could find the "body" node first (based on `name`) and then just take the first token before that node.